### ISSUE TITLE
[feature] Expose xdg-config/fontconfig to sandbox by default

### ISSUE THREAD

**Issue by yump** [flatpak/flatpak#3947]
## Linux distribution and version
Fedora 32

## Flatpak version
1.8.2

## Description of the problem

I've been trying to switch to flatpak Firefox, but I ran into weird and gross bitmap text on [some webpages](https://www.lesswrong.com/posts/R4f4RdGBdZsPzyJYk/a-discussion-of-heroic-responsibility#comments) and [Google sheets](https://docs.google.com/spreadsheets/d/1Lh6rHph31fiuijaZA0cff-9-LpVs9I-0SMm4uYuioCw/edit).  

Eventually, the problem was traced to [embedded bitmaps in some Microsoft fonts](https://ask.libreoffice.org/en/question/7127/calibri-font-not-rendering-correctly/).  It turns out that it was incidentally fixed for non-flatpak Firefox by the `~/.config/fontconfig/conf.d` fragment I use to disable hinting globally:

```
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
  <match target="pattern">
    <edit name="hinting" mode="assign"><bool>true</bool></edit>
    <edit name="hintstyle" mode="assign"><const>hintnone</const></edit>
  </match>
</fontconfig>
```

## Steps to reproduce

1. Have the Calibri font available, either copied to `~/.local/share/fonts` from a Windows installation or installed through your package manager (probably in `lpf-cleartype-fonts` from RPMFusion).

2. Visit either a [web page](https://www.lesswrong.com/posts/R4f4RdGBdZsPzyJYk/a-discussion-of-heroic-responsibility#comments) or [Google sheet](https://docs.google.com/spreadsheets/d/1Lh6rHph31fiuijaZA0cff-9-LpVs9I-0SMm4uYuioCw/edit) that uses the Calibri font.

3. See text that looks like [this](https://imgur.com/a/9OGdW5Y)

## Local Fix

Adding `xdg-config/fontconfig:ro` to filesystem overrides with Flatseal resolved the problem on my machine.

## Additional information

According to the discussion in #3410, disabling hinting with xsettings or GSettings would get passed into the sandbox.  Unforuntately, these methods do not allow re-enabling hinting for specific fonts or specific applications.  I like hinting for code but hate it for prose.  Also, it looks like you would need a fontconfig snippet to Properly Fix the embedded bitmap problem while leaving hinting enabled.

This issue is similar to #1563, but applies to user fontconfig settings, rather than system-wide.  I think user settings are less likely to run into the compat and backcompat difficulty brought up there, because they are usually simpler and usually don't modify the font search path (because a user could just put fonts under `~/.local/share/fonts`).

**Comment by soredake** [comment]
https://github.com/flatpak/flatpak/issues/3007

**Comment by soredake** [comment]
On my monitor for example i need to set hint style to 'medium' for fonts to look good, so i vote for this.

**Comment by ishworgurung** [comment]
> Adding xdg-config/fontconfig:ro to filesystem overrides with Flatseal resolved the problem on my machine.

Thanks for this @yump. Although, not directly related, `Flatseal` was _immensely_ helpful and set me up in the right direction.

**Comment by nanonyme** [comment]
IMHO this is probably not correct as a generic solution though it's a good workaround. @alexlarsson any ideas?

**Comment by alexlarsson** [comment]
Fontconfig is problematic in general, because the configuration format is so complex (likely turing complete) and can do so many things with a weird dependency model where just just stack a bunch of code snippets from various files. The configure files available on the host will be based on the base configuration files on the distro fonts, whereas in the sandbox they will be based on the runtime fonts/config. 

I don't think its realistinc to just assume any fontconfig files from the host will work in the runtime. However, I'm aware that this is a problem that we wil have to solve. It just has never risen to the top of the todo list, partly because of how incredibly hairy it is.

**Comment by mcatanzaro** [comment]
I have a strawman proposal.

 * Assumption: the way to minimize user complaints is to adhere as closely as possible to host font configuration. Any deviation in the runtime is going to be considered undesirable by users who expect fonts in flatpak to look like fonts on the host.
 * Strawman proposal: always prefer host font configuration. Assume that host systems have reasonable fonts installed, and maybe even tell runtimes to stop shipping fonts. Apps that hard depend on specific fonts must bundle them and ship their own fontconfig snippets.
 * Document a subset of supported fontconfig filesystem locations that will be mounted in the sandbox. [This is what WebKit does, and nobody seems to be complaining.](https://trac.webkit.org/browser/webkit/trunk/Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp?rev=268472#L398) Any host fontconfig files that reference unsupported locations will break. If the host does not contain a configuration file at some supported location, consider masking it in the runtime (by mounting an empty file?) to guarantee that host configuration can never get merged with runtime configuration.

I think this might work because it's simple. Make no attempt to parse fontconfig files, and certainly don't attempt to "merge" them. That just seems too hard.

**Comment by yump** [comment]
I like the strawman proposal, although I think runtimes shipping *fonts* is probably much less problematic than shipping fontconfig snippets.

Shortly after reporting this issue, I spent a few hours trying to use fontconfig to make my browser fingerprint match a bone-stock installation of my distro, and I doubt anyone other than the original authors of fontconfig understand it well enough to even attempt referencing snippets outside the standard locations.  I'd rather write an Xorg.conf.

**Comment by nanonyme** [comment]
Note that runtime *must* ship at least one fontconfig snippet. The configuration that allows apps to bundle fonts and add their own fontconfig snippets is in the runtime fontconfig snippet.

**Comment by nanonyme** [comment]
https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/master/files/fontconfig-flatpak.conf is not horrifyingly complex

**Comment by soredake** [comment]
https://github.com/snapcore/snapcraft/blob/master/extensions/desktop/common/fonts

**Comment by alexlarsson** [comment]
In general, fontconfig config files are not backwards compat, and several times new versions of fontconfig has added stuff to the default config that older libfontconfig version died on. So, we can't realistically use the host config.

**Comment by YaLTeR** [comment]
Looks like user font and font cache directories are already mounted under `/run/host/user-fonts` and `/run/host/user-fonts-cache` here: https://github.com/flatpak/flatpak/blob/f552d9a3972eb3cabdebbd35b18f8025bcd88377/common/flatpak-run.c#L2092

Perhaps some user fontconfig directory (e.g. `~/.config/fontconfig`) could be mounted similarly, then included from [`fontconfig-flatpak.conf`](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/1f131d97f83eed40c6c79ca465cc8bd3f51ff48b/files/fontconfig-flatpak.conf)?

> In general, fontconfig config files are not backwards compat, and several times new versions of fontconfig has added stuff to the default config that older libfontconfig version died on. So, we can't realistically use the host config.

How bad is this in practice for user fontconfig files? Maybe what would work is using a separate folder (`~/.config/fontconfig-flatpak` or something) so people can put only simple changes there?

**Comment by Eonfge** [comment]
Could this be something to resolve in the Runtime instead? I made an issue with the Freedesktop-SDK to provide some fallbacks for common fonts so that many existing font-issues can be remedied:

https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1274

**Comment by aleixpol** [comment]
This has been brought up as a problem in KDE as well:

https://bugs.kde.org/show_bug.cgi?id=436864
https://invent.kde.org/packaging/flatpak-kde-runtime/-/merge_requests/52

**Comment by alois31** [comment]
> In general, fontconfig config files are not backwards compat, and several times new versions of fontconfig has added stuff to the default config that older libfontconfig version died on. So, we can't realistically use the host config.

What is the situation you have in mind where libfonconfig "died"?
* What exactly is the "default config"? The forest that's shipped in `/etc` or the snippet generated by DE tools in `$XDG_CONFIG_HOME/fontconfig/fonts.conf`?
* In what way did libfontconfig "die"? Throw a couple of warnings and load the rest of the file properly, ignore the file completely after throwing a bunch of warnings, missing fonts or outright crashing applications?

I have tried copying my `fonts.conf` snippet to various systems with old fontconfig versions as well as corrupting it in various ways to simulate backwards-incompatible changes. The worst I have achieved so far is to get fontconfig to ignore the file after throwing some warnings.

**Comment by ljrk0** [comment]
So from what I can gather:

1. If an app uses a portal to query e.g., gsettings for fonts;  they work fine.
2. If the app uses fontconfig, this isn't supported (as of now).

It seems that there's not much drive behind fixing (2), maybe because most GTK+/GNOME apps are unaffected.  However, some apps do not query gsettings/portal but use fontconfig, for various reasons:

1. They don't use GTK+/GNOME for display
2. They allow the user to select a custom display font

The (2) point includes many tools, such as IDEs / text editors (VS Code, Ghidra, ...) but also word processors (LibreOffice) or graphic design tools (Inkscape, GIMP).  It's *quite* unexpected to have sans-serif not mapping to the default sans-serif font of the system, e.g., Cantarell for GNOME.

Perhaps, a proxy fontconfig library could be built that exposes the same interface as fontconfig does, but for a selected number of queries (namely default font family & aliasing settings) in turn asks the portal, and otherwise falls back to actual fontconfig.

Otherwise I don't see a solution to this quite nagging problem.

**Comment by YaLTeR** [comment]
> most GTK+/GNOME apps are unaffected

FWIW anything using `GtkSourceView` is affected because that uses `monospace` from fontconfig.

**Comment by mcatanzaro** [comment]
> If the app uses fontconfig, this isn't supported (as of now).

It should work fine as long as you understand it's necessarily going to use fontconfig settings from the container, not the host system. Changing that seems hard.

> Perhaps, a proxy fontconfig library could be built that exposes the same interface as fontconfig does, but for a selected number of queries (namely default font family & aliasing settings) in turn asks the portal, and otherwise falls back to actual fontconfig.

It would probably be easier to change fontconfig itself, but yeah something along those lines could work.

**Comment by ljrk0** [comment]
> It should work fine as long as you understand it's necessarily going to use fontconfig settings from the container, not the host system. Changing that seems hard.

Sure, but that's far from what we should expect users to need to do.  I appreciate the possibility, but I think OOTB the same font settings (defaults) that work outside Flatpak should be honored inside :)

**Comment by RalfJung** [comment]
I also had trouble with bad font rendering inside the container for vscodium that was fixed by a `xdg-config/fontconfig:ro` override.

**Comment by mmBesar** [comment]
> I also had trouble with bad font rendering inside the container for vscodium that was fixed by a `xdg-config/fontconfig:ro` override.

Thanks, This is the way!

**Comment by mcatanzaro** [comment]
Just for avoidance of doubt, to save trouble for anyone who finds this issue in the future: if you do that, your fonts _will_ eventually break. Fontconfig configuration is not stable and host config cannot safely be used inside the sandbox. That's why this issue report is still open. It might work today, but you have no guarantees for tomorrow. Voids your warranty.

**Comment by yump** [comment]
First, since the workaround is for per-user fontconfigs in `$HOME`, which are manually maintained by the user, if it breaks inside flatpak it will also be broken on the host.  My own fontconfig is a very tiny snippet that sets the RGB subpixel order and filter, and historically almost all breakages have been caused not by failure to parse, but by downstream consumers deciding to get their settings from someplace else.

Second, if a flatpak doesn't pick up host fontconfig, [fonts are in some sense already broken](https://github.com/flatpak/flatpak/issues/4571#issuecomment-1345502025).

**Comment by mcatanzaro** [comment]
> First, since the workaround is for per-user fontconfigs in $HOME, which are manually maintained by the user, if it breaks inside flatpak it will also be broken on the host. 

No, because the version of fontconfig in your flatpak runtime is not expected to match the version of fontconfig on your host system.

**Comment by yump** [comment]
The versions don't match, but they are expected to both move forward.  If the config works in the host and in flatpak on the the day the user adds the overrride, then it should keep working until (unless) fontconfig breaks compatibility with existing user configs.  Which will also hit the host, possibly before or after flatpak depending on distribution cadence, so the user will have to sort it out regardless.

The only exception that comes to mind is newly installing a flatpak that uses a fontconfig version older than the one that the user wrote their config against, which becomes less and less likely over time.

**Comment by bam80** [comment]
> The only exception that comes to mind is newly installing a flatpak that uses a fontconfig version older than the one that the user wrote their config against, which becomes less and less likely over time.

Agree. In my impression, runtime's generally ship stuff more fresh than host's ones, and update oftener.
Taking it into consideration, we should barely hit any problems in the reality.

@alexlarsson @mcatanzaro should we reconsider then?

**Comment by mcatanzaro** [comment]
No, because that's not a valid assumption. In fact, the opposite is generally true for Fedora users, Arch users, etc.

**Comment by mcatanzaro** [comment]
I mean, when a Fontconfig update is released, Arch will update immediately. Fedora will generally update sooner, but will certainly update on at least a 6 month cadence. freedesktop-sdk may or may not update, but will certainly update on a 12 month cadence. So the cadence of the most commonly-used Flatpak runtime is slower than the cadence of common distros.

Unfortunately this is a really tough bug and nobody knows what to do with it. The best proposal I've seen thus far is:

> Perhaps, a proxy fontconfig library could be built that exposes the same interface as fontconfig does, but for a selected number of queries (namely default font family & aliasing settings) in turn asks the portal, and otherwise falls back to actual fontconfig.

**Comment by yump** [comment]
It becomes less likely over time because the user's fontconfig is written once against the system fontconfig and then effectively frozen. My `~/.config/fontconfig/conf.d` is [backwards-compatible all the way back to 2012](https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commits/main/conf.d/50-user.conf), and I have no reason to believe this is unusual.

**Comment by mcatanzaro** [comment]
Oh sorry, I see you're proposing to allow only the _user_ config from the host, and not also the system config. I guess that's arguably OK? It will probably result in serious bugs for users with extensive custom configuration, but we can say argue that users will be responsible for any breakage.

Users like to copy/paste configuration they find on the internet, though. I fear we'll likely get bug reports from disappointed users when it starts applying to flatpaks and doesn't work there. Maybe that's an acceptable trade-off.

**Comment by ilya-fedin** [comment]
Is monospace font a solved thing? I inspected a few GTK apps and none seem to read host monospace setting for fonts. They do one of two things: use fontconfig's `monospace` font face which reads fontconfig default which doesn't get forwarded to flatpak (example of apps: fractal, epiphany) or read `org.gnome.desktop.interface.monospace-font-name` gsettings key which doesn't work either as apparently gsettings doesn't handle such settings as portalized automatically (example of apps: gedit, geary). In case of gedit, it even prints the value it gets in its settings dialog and it's always Source Code Pro 10 which surely not the value I have in dconf, fontconfig or anywhere else on the host (but matches the default value of the key). It was surprising to see this isn't unified on toolkit level like it's in Qt's case.

In case of KDE apps, monospace font setting is provided but it relies on the ability to read kdeglobals... Well, at least it works (in KDE session).

**Comment by bam80** [comment]
> Oh sorry, I see you're proposing to allow only the _user_ config from the host, and not also the system config. I guess that's arguably OK?

Is this OK? Can we move forward, after all?
